Mio buffer writer

ywkaras opened the pull request:
better this time

Commit:
Add milestones for SSL Handshake timings

Commit:
Test two continuations for HTTP_*_CLOSE

Commit:
Adding plugin to test transaction/session order

Commit:
Add support for Forwarded HTTP header tag (RFC7239).

calavera committed:
Preserve Accept-Encoding header for H2 Server Push promise.

This change maintains API compatibility.

Preserve the Accept-Encoding header from the original client request to
send H2 Server Push promises. With this change, responses initiated by a
Server Push promise can be compressed based on the accepted encoding.

Otherwise, the response body can never be compressed.

Signed-off-by: David Calavera <david.calavera@gmail.com>

Commit:
Replace proxy.config.http.normalize_ae_gzip with normalize_ae w/ Brotli support Issue # 2100

persiaAziz-zz committed:
YTSATS-1464: Support set-redirect with READ_REQUEST_PRE_REMAP_HOOK

SolidWallOfCode committed:
Re-arrange ink_platform.h. (Issue #2444)
1) Move unconditional includes to top of file.
2) Remove duplicates.

zwoop committed:
Add V=1 to the make check builds

PSUdaemon committed:
Coverity CID #1224619 Uninitialized members

Commit:
Add MIOBufferWriter class which writes to an MIOBuffer instance.

SolidWallOfCode committed:
Doc: Minor fixups for duplicated milestones and stat typo.

SolidWallOfCode commented on iocore/eventsystem/I_MIOBufferWriter.h:
```diff
@@ -0,0 +1,138 @@
+  MIOBufferWriter &
+  write(const void *data_, size_t length) override
+  {
+    const char *data = reinterpret_cast<const char *>(data_);
```
`static_cast` doesn't work?

ywkaras replied:
Probably.  I generally like to use the scariest cast possible when converting from a void pointer.  But converting to char is not scary really, so static_cast would be fine.

SolidWallOfCode commented on iocore/eventsystem/I_MIOBufferWriter.h:
```diff
@@ -0,0 +1,138 @@
+  // Write the first n characters that have been placed in the auxiliary buffer.  This call invalidates the auxiliary buffer.
+  // This function should not be called if no auxiliary buffer is available.
+  //
+  virtual MIOBufferWriter &
```
Is `virtual` correct here? Maybe `override`? I don't think we'll have a subclass of this class.

ywkaras replied:
Sorry brainfart.

nqyy committed:
Runroot for traffic_layout

ywkaras commented:
Does it look like I'm using the write interface to MIOBuffer correctly?

SolidWallOfCode commented:
Looking at it, actually not. You should be using `start()` instead of `buf()`. The former returns the first byte of not yet written space, while the latter is the buffer itself (which may already have data in it). So more like
```
int64_t avail = _miob->block_write_avail();
if (0 == avail) { _miob->add_block(); avail = _miob->block_write_avail(); }
int64_t to_write = std::min(length, avail);
memcpy(_miob->start(), buff, to_write);
length -= to_write;
```

SolidWallOfCode commented:
Heh. One of the reasons I wanted this is because generically writing data to an MIOBuffer isn't obvious. It should be easy to say "I have a buffer of length N, write it all to this MIOBuffer".

ywkaras commented:
OK, you're saying I should _not_ be getting a pointer to the first writable buffer then.

SolidWallOfCode commented:
Yes. Because the first writeable buffer may not be empty.

ywkaras commented:
So you assigned the person you doesn't know the non-obvious interface?

SolidWallOfCode commented:
Nobody knows it, really.

SolidWallOfCode commented:
To clarify a bit, there are several pointers to the first writable buffer.
* buf -> start of the buffer.
* start -> start of empty (unwritten) space in the buffer.
* end -> one past the end of the buffer

SolidWallOfCode committed:
Cleanup: remove ts::ConstBuffer from ink_inet.
This turns out to have a lot of tentacles all over.

shukitchan committed:
add new TLS milestones

SolidWallOfCode committed:
Cleanup: Fix traffic_manager so its checks run if WCCP is enabled.

ywkaras commented:
@SolidWallOfCode do you mean the member functions of IOBufferBlock in I_IOBuffer.h ?  You seem to be contradicting the code comments.  They say start() is the beginning of the "Inuse" section, which I assume means the already-written data.  They say end() is the end of the used space.  I assumed this was "end" in the STL sense, one past the last byte, the place to start adding more data.  It seems like one could possibly use member functions of MIOBuffer, which you seemed to suggest a few messages ago.  I'm going to try looking for examples of use in the code, and will perhaps die of shock if I find consistent usage.

ywkaras commented:
Looking at iocore/net/SSLNetVConnection.cc , it would appear that my code is using the interface properly.

SolidWallOfCode commented:
Yes, looks like you're correct. Oh well.

shukitchan committed:
Update document

SolidWallOfCode committed:
Add string literal support to ts::string_view.

SolidWallOfCode committed:
Doc: Clean up comments in null_transform example plugin.

SolidWallOfCode committed:
Tests: Clean up file extensions in lib/ts unit tests.

SolidWallOfCode committed:
Test: Fix timing issues with txn and double tests.
Also tweak test_UDPNet to report errors better and use a less common port.

bryancall committed:
Update s3auth test, removed cpp file and added the Catch main define in the
unit test file

bryancall committed:
Ignoring unit tests

oknet committed:
Support dynamic registration to StatPages.

The cache stat module will register to StatPages after cache system
initialized. But there is not a mutex to protect the stat_pages[] array.
It will leads a rare crash when proxy.config.http.wait_for_cache is
false.

zwoop committed:
Make sure all files in the sandbox are readable

SolidWallOfCode committed:
BufferWriter: Add data() method for access to the internal buffer for API consistency with std:string and ts::string_view.

SolidWallOfCode committed:
Cleanup: fix rollback to use elevating file operations.

SolidWallOfCode committed:
ESI: Make maximum file count runtime configurable.

SolidWallOfCode committed:
Fix runtime undefined symbol error in multiplexer plugin.

SolidWallOfCode committed:
UDPNet: Yet another timing fix for test_UDPnet.

SolidWallOfCode committed:
Doc: Tweak makefiles to not build man pages for install if doc building is disabled.

PSUdaemon committed:
Remove volatile.

bryancall committed:
Added std::make_unique for C++11

(cherry picked from commit c4c3bd9c213b63d03108653761be90f52ac55c33)

jhg03a committed:
Fixed typo in cachekey plugin documentation

shukitchan committed:
Fix mem leak related to do_os_response() function

zwoop committed:
Change the include from .cc to .h, as Odin intended

PSUdaemon committed:
Update .gitignore
